include/zephyr/scsi/scsi.h

```diff
@@ -0,0 +1,210 @@
+#define SG_DXFER_NONE		(1)
+#define SG_DXFER_TO_DEV		(2)
+#define SG_DXFER_FROM_DEV	(3)
+
+/* STRUCTURES */
+
+/**
+ * @brief Structure to hold information about a SCSI command to be processed
+ */
+struct scsi_cmd {
+	uint8_t cmd[MAX_CDB_SIZE]; /**< Command descriptor block (CDB) */
+	uint8_t lun;               /**< Target LUN (Logical Unit Number) */
+	uint8_t cmdlen;            /**< Length of the command */
+	uint64_t datalen;          /**< Total length of data to be transferred */
+	uint8_t *pdata;            /**< Pointer to data to be transferred */
+	uint8_t dma_dir;           /**< Direction of data transfer (e.g., read/write) */
+} __packed;
+
+/**
+ * @brief Structure representing LUN (Logical Unit Number) information.
+ */
+struct lun_info {
+	uint32_t lun_id;      /**< ID of the LUN */
+	bool lun_enabled;     /**< Indicates if the LUN is enabled */
+	uint32_t block_size;  /**< Size of each block in bytes */
+	uint64_t block_count; /**< Total number of blocks in the LUN */
+} __packed;
+
+/**
+ * @brief Information about a SCSI host controller.
+ */
+struct scsi_host_info {
+	sys_slist_t sdevices;          /**< List of connected SCSI devices */
+	const struct scsi_ops *ops;    /**< Pointer to the SCSI operations structure */
+	void *hostdata;                /**< Pointer to host-specific data */
+	struct device *parent;         /**< Parent device associated with the host */
+};
+
+/**
+ * @brief Struct representing a SCSI device
+ */
+struct scsi_device {
+	sys_snode_t node;            /**< Node in the list of SCSI devices */
+	struct scsi_host_info *host; /**< Pointer to the associated SCSI host controller */
+	uint8_t lun;                 /**< LUN ID for the device */
+	uint32_t sector_size;        /**< Size of sectors in bytes */
+	uint32_t capacity;           /**< Total capacity of the device in blocks */
+};
+
+/**
+ * @brief Structure to represent a SCSI Generic (SG) I/O request
+ */
+struct sg_io_req {
+	uint32_t protocol;         /**< SCSI protocol type */
+	uint32_t subprotocol;      /**< Sub-protocol type (0 - SCSI command, 1 - SCSI transport) */
+	void *request;             /**< Pointer to the SCSI CDB or transport layer request */
+	uint32_t request_len;      /**< Length of the request in bytes */
+	void *response;            /**< Pointer to the response buffer */
+	uint32_t max_response_len; /**< Maximum length of the response buffer */
+	int32_t dxfer_dir;         /**< Data transfer direction */
+	uint32_t dxfer_len;        /**< Length of the data to transfer */
+	void *dxferp;              /**< Pointer to the data to be transferred */
+};
+
+/**
+ * @brief Operations for managing SCSI commands
+ */
+struct scsi_ops {
+	/**
+	 * @brief Execute a SCSI command on a given SCSI device.
+	 *
+	 * @param sdev Pointer to the SCSI device handle
+	 * @param cmd Pointer to the SCSI command to execute
+	 *
+	 * @return 0 on success, negative value on error
+	 */
+	int32_t (*exec)(struct scsi_device *sdev, struct scsi_cmd *cmd);
```

Comment:
Should `scsi` be a driver class or a subsystem? I guess the difference here is that there is a lot of generic code for SCSI, but we still need specific drivers to handle SCSI commands?